sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ObjectExpressionsSuite.scala

```diff
@@ -423,7 +423,7 @@ class ObjectExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
       inputTypes = Nil,
       propagateNull = false,
       dataType = ObjectType(classOf[ScroogeLikeExample]),
-      outerPointer = Some(() => outerObj))
+      outerPointer = None)
```

Comment:
Shall we create a new test case?

Reply:
I am not sure I understand what the new and the old test case would be. My understanding is that `outerPointer` is only for inner classes but `ScroogeLikeExample` is not an inner class. So the old test case was incorrect and only works because the old code dropped the outer param by incorrectly calling tail on the param list. So this change is necessary for this spec not to fail.

Reply:
So normal inner classes (that uses outer pointer) are tested here in the ExpressionEncoderSuite: https://github.com/apache/spark/blob/ae08787f5c50e485ef4432a0c2da8b3b7290d725/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoderSuite.scala#L229-L233
and here in the ObjectExpressionsSuite: https://github.com/apache/spark/blob/ae08787f5c50e485ef4432a0c2da8b3b7290d725/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ObjectExpressionsSuite.scala#L408-L417
so such cases are covered.

Are you saying that we should add a test case for a class that only have an apply side constructor? Trying to do something like that will not work in either this branch or master. This is because companion objects of inner classes are not singletons and the codegen will fail with that `"MODULE$" is neither a method, a field` because of this. Such a class would also behave slightly differently as the apply method constructor would not take an outerPointer. This is because the companion object already has an outer pointer and that will be used when creating the inner class object. Maybe it would be possible to add support for such cases but it would require more changes and is probably out of scope for this PR.

But just to be clear *both* the test and the code was wrong before this PR and they were wrong in such a way were they cancelled out. And the new spec in ExpressionEncoderSuite in this PR that tests at a "higher level" shows also that the previous code was wrong as that test case will fail on master.

Reply:
@HyukjinKwon Does my comment make sense? Would be great if this could make into 3.3.1 before rc2.